src/Xamarin.Android.Build.Tasks/Utilities/NativeAssemblyGenerator/NativeAssemblyGenerator.cs

```diff
@@ -0,0 +1,1223 @@
+
+		public void WriteEOL (string? comment = null, bool useBlockComment = false)
+		{
+			WriteEOL (Output, comment, useBlockComment);
+		}
+
+		public void WriteEOL (TextWriter writer, string? comment = null, bool useBlockComment = false)
+		{
+			if (!String.IsNullOrEmpty (comment)) {
+				WriteCommentLine (writer, comment, useBlockComment: useBlockComment);
+				return;
+			}
+			writer.WriteLine ();
+		}
+
+		public void WriteInclude (string filePath, string? comment = null, bool useBlockComment = false)
+		{
+			WriteInclude (Output, filePath, comment, useBlockComment);
+		}
+
+		public void WriteInclude (TextWriter writer, string filePath, string? comment = null, bool useBlockComment = false)
+		{
+			WriteDirectiveWithComment (writer, ".include", comment, useBlockComment, QuoteString (filePath));
+		}
+
+		public void WriteCommentLine (string? comment = null, bool indent = true, bool useBlockComment = false)
+		{
+			WriteCommentLine (Output, comment, indent);
+		}
+
+		public void WriteCommentLine (TextWriter writer, string? comment = null, bool indent = true, bool useBlockComment = false)
+		{
+			if (indent) {
+				writer.Write (Indent);
+			}
+
+			if (useBlockComment) {
+				writer.Write ("/* ");
+			} else {
+				writer.Write (LineCommentStart);
+			}
+
+			if (!String.IsNullOrEmpty (comment)) {
+				writer.Write (' ');
+				writer.Write (comment);
+			}
+
+			if (useBlockComment) {
+				writer.Write (" */");
+			}
+
+			writer.WriteLine ();
+		}
+
+		public StructureWriteContext StartStructure () => new StructureWriteContext ();
+		public StructureWriteContext StartStructureArray () => new StructureWriteContext (isArray: true);
+
+		List<StructureMemberInfo> GatherMembers<T> () where T: class
+		{
+			Type klass = typeof(T);
+			var ret = new List<StructureMemberInfo> ();
+
+			foreach (MemberInfo mi in klass.GetMembers ()) {
+				Type? memberType = mi switch {
+					FieldInfo fi => fi.FieldType,
+					PropertyInfo pi => GetPropertyType (pi),
+					_ => null
+				};
+
+				if (memberType != null) {
+					NativeAssemblerAttribute? attr = mi.GetCustomAttribute <NativeAssemblerAttribute> ();
+					if (attr == null || !attr.Ignore) {
+						ret.Add (new StructureMemberInfo (mi, memberType, attr));
+					}
+				}
+			}
+
+			return ret;
+
+			Type GetPropertyType (PropertyInfo pi)
+			{
+				if (!pi.CanRead) {
+					throw new InvalidOperationException ($"Property '{pi.DeclaringType}.{pi.Name}' doesn't have a getter.");
+				}
+
+				return pi.PropertyType;
+			}
+		}
+
+		public StructureWriteContext AddStructureArrayElement (StructureWriteContext status)
+		{
+			if (status.ArrayOfStructures == null) {
+				throw new InvalidOperationException ("Structure context doesn't refer to array of structures, did you use StartStructureArray()?");
+			}
+
+			var ret = StartStructure ();
+			status.ArrayOfStructures.Add (ret);
+
+			return ret;
+		}
+
+		public void WriteStructure<T> (StructureWriteContext status, T[] collection) where T: class
```

Comment:
I find it slightly amusing that a "structure" must be a `class`, but I can't think of a better name either…